### ISSUE TITLE
GPL v3.0 License Choice

### ISSUE THREAD

**Issue by MichealReed** [freqtrade/freqtrade#1827]
As I understand it, under the GPL 3.0 license, freqtrade could be used in a closed source project as long as the product is not distributed externally. If distribution happens, the binaries must include a copy of the source. I am concerned that this could be interpreted to include custom indicators, signals, and strategies as well. Am I interpreting this correctly?

**Comment by hroff-1902** [comment]
Think so. GPL is viral. It has viral-spread of the GPL conditions back to your own code, distributed alnong with even small GPLed parts.

However, how can you imagine distribution of your freqtrade instance with your "custom indicators, signals, and strategies" without copies of sources taking into account that all this stuff are python texts? They will be in the distribution in any case. If the question is: should they be GPLed as well, the answer is: I guess no (just imho, discussible). But sources of all parts including your custom parts should be available anyway. In case of python code it's completed automatically.

Just imho.

**Comment by MichealReed** [comment]
There are methods via compiled python, obfuscation, and creating interfaces with compiled lower level libraries to not disclose code. In this situation, the majority of the source code would be distributed (and potentially contributed back here if the improvement is needed publicly), but the above outlined proprietary components would not. 

Given this, although I ideologically agree with open source, I'm not sure GPL v3.0 is the best fit for a trading bot. Maybe our purpose is a very edge case, but there may be others infringing without knowledge.

**Comment by hroff-1902** [comment]
> There are methods via compiled python, obfuscation, and creating interfaces with compiled lower level libraries to not disclose code. In this situation, the majority of the source code would be distributed (and potentially contributed back here if the improvement is needed publicly), but the above outlined proprietary components would not.

Well, all this (compiled python, obfuscation, and creating interfaces with compiled lower level libraries) is not a part of freqtrade and not supported by freqtrade team, as far as I know, It will be a question of your distribution if you apply any of those techniques, how to comply to GPL requirements...

> Given this, although I ideologically agree with open source, I'm not sure GPL v3.0 is the best fit for a trading bot. Maybe our purpose is a very edge case, but there may be others infringing without knowledge.

Here I agree with you. A less restrictive license would be better (like BSD or Apache, what else?) for end-users. Let's ask @xmatthias and the team to consider this point...

**Comment by xmatthias** [comment]
The license has been in place when i found this bot - I'm not certain on the choice - but i **guess** it's to protect the bot from having a corporation come in, grab the code, build a closed-source application out of it and sell it (at least that's my take on GPL).


maybe @gcarq as project initiator can give us some insight to the license decision?

**Comment by hroff-1902** [comment]
Ah, yeah -- we need permission from all contrubutors to relicense project they contributed to. Just forget it...

**Comment by gcarq** [comment]
I'm no legal expert and choosing GPL seemed to be the best fit for the project, very much for the reasons pointed out by @xmatthias. 

I don't think custom indicators, strategies and everything outside of the tree strictly need the same license. As an example the linux kernel is licensed with GPLv2 with some firmware images released under non-free licenses.

**Comment by hroff-1902** [comment]
Hi @gcarq!

As far as I understand (and remember) the situation with non-free firmware blobs in linux -- they are separate programs and not parts of gpled kernel distribution...

The problem here is that the GPL license is viral. Strictly speaking I (and everyone else) is allowed to request sources of the custom strategies from every user complaining that she has a problem running freqtrade with their custom strategies/hyperopts instead of default included in the freqtrade distribution since they run that code as a single program with GPLed freqtrade...

But changing the license is not really achievable -- it requires permissions from every contributor.

Just imho, I'm also not a legal expert...

**Comment by gcarq** [comment]
@hroff-1902 I don't think the GPL license states that I have the right to get a copy of your custom strategy, just because I know you have one (no matter which license). GPL only applies if you want to distribute freqtrade (e.g.: selling freqtrade directly to a customer along your non-free trading strategy). You can still provide a paid service which incorporates freqtrade and your custom strategy.


Here is a link to the GPL FAQ about this aggregation https://www.gnu.org/licenses/gpl-faq.en.html#MereAggregation.

> An “aggregate” consists of a number of separate programs, distributed together on the same CD-ROM or other media. The GPL permits you to create and distribute an aggregate, even when the licenses of the other software are nonfree or GPL-incompatible. The only condition is that you cannot release the aggregate under a license that prohibits users from exercising rights that each program's individual license would grant them.

> Where's the line between two separate programs, and one program with two parts? This is a legal question, which ultimately judges will decide. We believe that a proper criterion depends both on the mechanism of communication (exec, pipes, rpc, function calls within a shared address space, etc.) and the semantics of the communication (what kinds of information are interchanged).

> If the modules are included in the same executable file, they are definitely combined in one program. If modules are designed to run linked together in a shared address space, that almost surely means combining them into one program.

> By contrast, pipes, sockets and command-line arguments are communication mechanisms normally used between two separate programs. So when they are used for communication, the modules normally are separate programs. But if the semantics of the communication are intimate enough, exchanging complex internal data structures, that too could be a basis to consider the two parts as combined into a larger program.

But yes, changing the license isn't really achievable as you mentioned.

**Comment by hroff-1902** [comment]
> GPL only applies if you want to distribute freqtrade (e.g.: selling freqtrade directly to a customer along your non-free trading strategy). You can still provide a paid service which incorporates freqtrade and your custom strategy.

Agree, I (wrongly) thought that GPL is applied just after creation of a new modified version of the program, straight after the fact of the creation of a new modified program. From the same document:

> Does the GPL require that source code of modified versions be posted to the public? 
>
> The GPL does not require you to release your modified version, or any part of it. You are free to make modifications and use them privately, without ever releasing them. This applies to organizations (including companies), too; an organization can make a modified version and use it internally without ever releasing it outside the organization.
>
>But if you release the modified version to the public in some way, the GPL requires you to make the modified source code available to the program's users, under the GPL.

Description about aggregation is not applied though for the custom licenses and hyperopts since these are python modules that are resolved by freqtrade and run in the context of freqtrade. I.e. its like linking, not aggregation of the different programs with different licenses on a media or other distribution.

Okay, in short, GPL is applied for distributed versions of freqtrade, In this case GPL is okay. By using GPL we require users to provide the sources only if they start distribution of the modified freqtrade.

**Comment by hroff-1902** [comment]
Closing it. Feel free to reopen if there will be some more points to discuss regarding the license and license transition.
